pkg/server/infrastructure/datastore/mysql/mysql.go

```diff
@@ -47,40 +64,249 @@ func New(ctx context.Context, cfg datastore.Config) (datastore.DataStore, error)
 func (m *mysql) BatchAdd(ctx context.Context, entities []datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	notRollback := make(map[string]bool)
+	for i, saveEntity := range entities {
+		if err := m.Add(ctx, saveEntity); err != nil {
+			if errors.Is(err, datastore.ErrRecordExist) {
+				notRollback[saveEntity.PrimaryKey()] = true
+			}
+			for _, deleteEntity := range entities[:i] {
+				if _, exit := notRollback[deleteEntity.PrimaryKey()]; !exit {
+					if err := m.Delete(ctx, deleteEntity); err != nil {
+						if !errors.Is(err, datastore.ErrRecordNotExist) {
+							klog.Errorf("rollback delete entity failure %w", err)
+						}
+					}
+				}
+			}
+			return datastore.NewDBError(fmt.Errorf("save entities occur error, %w", err))
+		}
+	}
+	return nil
 }
 
 // Get get data model
 func (m *mysql) Get(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+
+	if dbGet := m.client.First(entity); dbGet.Error != nil {
+		if errors.Is(dbGet.Error, gorm.ErrRecordNotFound) {
+			return datastore.ErrRecordNotExist
+		}
+		return datastore.NewDBError(dbGet.Error)
+	}
+	return nil
 }
 
 // Put update data model
 func (m *mysql) Put(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+	entity.SetUpdateTime(time.Now())
+	if dbPut := m.client.Model(entity).Updates(entity); dbPut.Error != nil {
+		if errors.Is(dbPut.Error, gorm.ErrRecordNotFound) {
+			return datastore.ErrRecordNotExist
+		}
+		return datastore.NewDBError(dbPut.Error)
+	}
+	return nil
 }
 
 // IsExist determine whether data exists.
 func (m *mysql) IsExist(ctx context.Context, entity datastore.Entity) (bool, error) {
-	return true, fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return false, datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return false, datastore.ErrTableNameEmpty
+	}
+
+	if dbExist := m.client.First(entity); dbExist.Error != nil {
+		if errors.Is(dbExist.Error, gorm.ErrRecordNotFound) {
+			return false, nil
+		}
+		return false, datastore.NewDBError(dbExist.Error)
+	}
+
+	return true, nil
 }
 
 // Delete delete data
 func (m *mysql) Delete(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+	// check entity is exist
+	if err := m.Get(ctx, entity); err != nil {
+		return err
+	}
+
+	if dbDelete := m.client.Unscoped().Model(entity).Delete(entity); dbDelete.Error != nil {
```

Comment:
Why we need Unscoped here? We are not performing soft deleting in any model so I think it's OK to just delete the entity.

Reply:
Yeah, correct. As we don't have `gorm.DeletedAt` field in our model, `gorm.Delete()` will not perform soft delete. But if we would have the `gorm.DeletedAt` field in our model, then `db.Unscoped().Delete()` would have deleted the record permanently according to the gorm [documentation](https://gorm.io/docs/delete.html#Delete-permanently).

Reply:
Done